### PR TITLE
Add opt-in rowwise-only quantized primary weights

### PR DESCRIPTION
**Pull request by xiuhu17**
# Description

`quantized_model_init` normally allocates both rowwise and columnwise storage for primary quantized weights when autograd is enabled. The columnwise representation is required by the default quantized dgrad, but it is not consumed when a fixed `backward_override="high_precision"` or `"dequantized"` mode computes dgrad from a high-precision or dequantized weight.

This matters for LoRA-style training with a large frozen quantized base: the base still participates in dgrad, while an unused columnwise copy remains attached to every primary weight.

This PR adds an explicit experimental opt-in:

```python
with te.quantized_model_init(
    enabled=True,
    recipe=recipe,
    omit_columnwise_primary_weight_storage=True,
):
    model = ...
```

When enabled, primary quantized weights allocate only the rowwise representation required by forward. The option requires an initialization recipe whose `backward_override` is `"high_precision"` or `"dequantized"`.

## Compatibility and safety

- The default is `False`, so existing callers retain bidirectional primary-weight storage, including callers that switch from a backward override to default quantized backward at runtime.
- Opt-in models fail early with a descriptive error if a later autocast recipe requests default quantized backward.
- The option is intended for fixed-mode workloads such as frozen LoRA base weights. Optimizer or distributed master-weight writeback paths that require columnwise primary storage must leave it disabled.

For block-scaled formats, omitting one persistent direction saves approximately 1.03 bytes/parameter for MXFP8 and 0.56 bytes/parameter for 1D NVFP4 before shape padding. The NVFP4 test explicitly uses one-dimensional scaling, which cannot synthesize columnwise storage from rowwise storage.

## Type of change

- [x] Documentation change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

- Add `omit_columnwise_primary_weight_storage` to `quantized_model_init` and preserve it across nested context state.
- Apply the opt-in when allocating primary-weight storage.
- Validate the initialization recipe and reject an incompatible runtime switch before a quantized dgrad requests missing storage.
- Add MXFP8 and 1D NVFP4 coverage for default storage, rowwise-only storage, complete override forward/backward, preserved default runtime switching, and fail-fast behavior.

## Validation

- Repository-configured Black 24.4.2 passed for all three changed Python files.
- Python syntax compilation passed for all three changed Python files.
- Pylint passed for both changed Transformer Engine source files (10.00/10).
- `git diff --check` passed.
- CUDA tests were not run locally because this host has no CUDA/Blackwell runtime; the added tests use the existing MXFP8/NVFP4 availability gates.
- Full recursive Python lint is blocked on this host by the repository-pinned Astroid version parsing Python 3.14 `TemplateStr`; the changed source files pass the pinned Pylint independently.

# Checklist

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the public API documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove the feature works and preserves default compatibility
- [ ] New and existing GPU unit tests pass locally (CUDA/Blackwell runtime unavailable locally)

### CODE REVIEW

File: transformer_engine/pytorch/module/base.py [hunk 1]
```diff
@@ -1845,7 +1845,15 @@ def reset_parameters(self, defer_init: Optional[bool] = False) -> None:
...
                     raise RuntimeError("Weight quantizer has not been initialized")
-                quantizer.set_usage(rowwise=True, columnwise=torch.is_grad_enabled())
+                # Backward overrides consume high-precision or dequantized weights,
+                # so they do not need a quantized columnwise representation.
+                quantizer.set_usage(
+                    rowwise=True,
+                    columnwise=(
+                        torch.is_grad_enabled()
+                        and self.fp8_meta["recipe"].backward_override is None
+                    ),
```
[the remark this answers is not in the document]
Comment (xiuhu17) [reply transformer_engine/pytorch/module/base.py:1855]: Fixed in bae76d17. Rowwise-only primary storage is now an explicit opt-in (`omit_columnwise_primary_weight_storage=True`) whose default is off, so existing models initialized under an override retain columnwise storage and can switch back to quantized backward. The opt-in validates the initialization recipe and fails early if a later autocast recipe requests quantized backward. Tests now cover the preserved default override-to-quantized switch for both MXFP8 and 1D NVFP4, plus the opt-in fail-fast contract.

**Comment by xiuhu17** [comment]
@ptrendx Would you like to take a look at it?
